### ISSUE TITLE
API functions for mesh connectivity are ambiguous

### ISSUE THREAD

**Issue by ajaust** [precice/precice#1057]
**Please describe the problem you are trying to solve.**
The API functions connected to setting the connectivity of triangles and quad meshes have ambiguous names as well as ambiguous names for the input parameters. The functions in question are the following

```c++
void setMeshTriangle(int meshID, int firstEdgeID, int secondEdgeID, int thirdEdgeID);
void setMeshTriangleWithEdges(int meshID, int firstVertexID, int secondVertexID, int thirdVertexID);

void setMeshQuad(int meshID, int firstEdgeID, int secondEdgeID, int thirdEdgeID, int fourthEdgeID);
void setMeshQuadWithEdges(int meshID, int firstVertexID, int secondVertexID, int thirdVertexID, int fourthVertexID);
```

1. Why do the functions `setMeshTriangleWithEdges` and `setMeshQuadWithEdges` explicitly refer to edges in their name, but do **not** expect edge ids? The name of these functions could be interpreted as setting a triangle/quad with the edges provided to the function. 
2. Why do you have one set of functions referring to the edges, i.e. the functions in 1., while the other functions `setMeshTriangle` and `setMeshQuad` do **not** explicitly refer to vertices?
3. From the wording it is not clear if the function sets some option of an existing triangle/quad or if it should create something new.
4. I see a potential risk since for bugs by mixing the functions up since these functions also have the same signature `(int, int, int, int)`.
5. Something I just realized now: Why are none of the parameters `const int`?

**Describe the solution you propose.**
I would propose a consistent naming scheme for all types of mesh element creation. I would propose something like this:

```c++
void createMeshTriangleFromEdges(const int meshID, const  int firstEdgeID, const int secondEdgeID, const int thirdEdgeID);
void createMeshTriangleFromVertices(const int meshID, const int firstVertexID, const int secondVertexID, const int thirdVertexID);

void createMeshQuadFromEdges(const int meshID, const int firstEdgeID, const int secondEdgeID, const int thirdEdgeID, const int fourthEdgeID);
void createMeshQuadFromVertices(const int meshID, const int firstVertexID, const int secondVertexID, const int thirdVertexID, const int fourthVertexID);
```
To me this has the following advantages:
1. It is clear whether the functions expect edge ids or vertex ids.
2. The naming scheme of the "edge" and the "vertex" version are consistent.
3. It is clear that it creates the logic of a connectivity between certain points/edges. Although I wound not be too strict here. If you like `setXXX` more than `createXXX` than use the former.
4. More explicit names on what the functions expect should minimize the risk of bugs by mixing up the functions.
5. Fixes the `const`ness of parameters.

**Describe alternatives you've considered**
None

**Additional context**
This issue would affect https://github.com/precice/precice/issues/465

**Comment by DavidSCN** [comment]
Thanks for pointing this out. I guess the naming scheme has historical reasons, we first had `setMeshEdges` and `setMeshTriangles` and only later on an alternative to `setMeshTriangles`, namely `setMeshTrianglesWithEdges`, which creates the edges on the fly. 

However, I completely agree on the more verbose and clear naming scheme you propose, we just need to be careful when deprecating and renaming API functions. Maybe it would be a good idea to already add the functions now and remove the 'old' functions completely for the next major release `v3.x.x`, so that we still have some backward compatibility.

> 5. Something I just realized now: Why are none of the parameters const int?

This is a question I frequently discuss with @fsimonis and I'm voting for `cost`'ness here as well :laughing:

**Comment by MakisH** [comment]
> > 1. Something I just realized now: Why are none of the parameters const int?
> 
> This is a question I frequently discuss with @fsimonis and I'm voting for `cost`'ness here as well laughing

Apart from any potential reasons to have specific things not `const` by design:
- isn't `const` part of the ABI? So, we should hold such a change for v3.0.
- potentially this has something to do with the interoperability with other languages?

**Comment by uekerman** [comment]
Yes, our current naming scheme is not perfect here. I always mix them up as well.

The reasoning for the current naming scheme is:

* `setMeshTriangle` creates a triangle
* `setMeshTriangleWithEdges` create a triangle while creating the necessary edges on the fly

Indeed, the "with" is very easy to misunderstand as _create triangles with the edges I am giving you_.

In the proposed naming scheme `createMeshTriangleFromVertices`, we lose the information that edges are created as well. Maybe this is OK since this is anyway pretty hidden from the user.

Let's improve the overall naming here with v3. In particular, "create" instead of "set" is much needed in my opinion. But then we also need to carry this over to `setMeshVertex" etc. Maybe we could even drop the "Mesh".

**Comment by fsimonis** [comment]
Resolved by #1322
